### PR TITLE
Make parameter shape tests deterministic.

### LINKED ISSUE #16: Review approach to fixing rng state for each test case
I copied the code for this verbatim from pyro in order get the tests working standalone.

* Tests were flakier than expected without this
* Assuming we continue to fix the seed, is there anything to do for numpyro?

### PR DESCRIPTION
**Pull request by null-a**
Prior to this change, the model built in this test depended in part on randomly generated data. To make this test deterministic it was necessary to fix the random seed, which happened in `conftest.py`.

(`dummy_df` generates  a random data frame from metadata describing the desired data set. e.g. the metadata might say the `y` ought to be real valued, and `x` ought to be categorical with levels `a` and `b`. This data frame is then used as part of the model definition, i.e. it is passed to `defm`. The coefficients present in the model depend on the levels of the categorical columns in the data, and since `dummy_df` may or may not sample all of the levels described by the metadata, the model returned by `defm` is no longer deterministic.)

Following this change, the model is built directly from the metadata that describes the data set. Since this is a deterministic process there is no longer a need to fix the rng seed.

This is the only test that relies on a fixed seed as far as I can tell.

Closes #16.

### CODE REVIEW

File: tests/test_brm.py [hunk 1]
```diff
@@ -436,13 +436,15 @@ def test_parameter_shapes(formula_str, non_real_cols, contrasts, family, priors,
     df = dummy_df(cols, N)
 
     # Define model, and generate a single posterior sample.
-    model = defm(formula_str, df, family, priors, contrasts)
-    fit = model.fit(**fitargs)
+    metadata = metadata_from_cols(cols)
+    desc = makedesc(formula, metadata, family, priors, code_lengths(contrasts))
+    data = makedata(formula, df, metadata, contrasts)
+    fit = DefmResult(formula, metadata, contrasts, desc, data).fit(**fitargs)
```
[the remark this answers is not in the document]
Comment (null-a) [reply tests/test_brm.py]: > the reason why this gives deterministic results now is because the metadata variable that is being passed in has all the information about the different levels and this is being inferred from dummy_df? 

That sounds like it might be the reason it was non-deterministic previously.

You're right that `metadata` carries information about which levels a particular categorical column can take. (Along with other info.) This information is part of the definition of each test case. (It comes via `non_real_cols`, but we can just think of `metadata` as coming from the test case definition for present purposes.)

Previously, this `metadata` was used to generate a `dummy_df` which was then handed to `defm` to set up the model. Internally, `defm` (re)extracts the metadata from the data frame it is given, using the information about e.g. categorical levels to decide how to encode the design matrix, and which coefficients are required in the model. This (re)extracting (by `defm`) of the metadata is problematic because it can lose information when compared to the original metadata in the test definition.

Following this change we still generate a data frame from the metadata given as part of the test definition using `dummy_df`. (It's fine that the data is random for this test. It only cares about the shape of things, and that don't depend on the data.) But rather than set the model up from the data frame (using `defm`) we instead use the `metadata` directly. (Using internal method `makedesc`.) This avoids the problematic lossy (re)extraction of the metadata, so the structure of the model built by the test no longer depends on the rng seed.

> Is dummy_df used only for testing purposes?

Within the brmp package, yes I think so. I've probably used it in notebooks elsewhere.

> If so, is it worth having an option that...

Yes, I went that way originally too, and have a half finished local branch with a version of `dummy_df` that's similar to what you describe. But in the end, I felt that building models directly from the metadata was a sensible approach to take for testing in general (the other code generation tests already work this way) and was sufficient to fix this problem. One concrete thing you can do with this approach that you can't do with the modified `dummy_df` approach is run some of these tests on data frames with very few rows. (e.g. `N=0` would always trip assertion.) This is something we make use of. (It's not at all obvious why it's useful to do so, but it's related to the OED branch. Happy to chat about more if you're interested.)

**Comment by null-a** [comment]
This is great, thanks.

> I would imagine that there would be many sources of randomness including the generated data, but I guess the level at which we are testing, e.g. code generation, this randomness does not play a role. Is that correct?

Exactly.

> Btw, feel free to merge PRs after review, as you see fit. That's the approach we have been taking in other repos like Funsor and NumPyro, i.e. authors are free to merge their PRs after approval.

Will do!
